FRP/Euphoria/Update.hs

```diff
@@ -58,59 +59,45 @@ import FRP.Euphoria.Event
 -- Note: in the current implementation, if you use an 'Update' twice,
 -- an unbounded amount of computation can be duplicated. Please
 -- avoid doing so.
-data Update a = forall s. (Monoid s) => Update (s -> a) (Event s)
+data Update a = forall s. (Semigroup s) => Update (s -> a) (Event s)
 
 instance Functor Update where
     f `fmap` Update final evt = Update (f . final) evt
 
-instance Applicative Update where
-    pure x = Update (const x) (mempty :: Event ())
-    Update f_final f_evt <*> Update a_final a_evt = Update
-        (\(f_s, a_s) -> f_final f_s (a_final a_s))
-        ((left <$> f_evt) `mappend` (right <$> a_evt))
-        where
-            left f = (f, mempty)
-            right a = (mempty, a)
+instance (Semigroup a) => Semigroup (Update a) where
+    Update f x <> Update g y = Update
+        (these f g (\a b -> f a <> g b))
+        (fmap This x <> fmap That y)
```

Comment:
I like the use of These in this instance.

Reply:
Thank you!